src/core/seat/input-method-relay.cpp

```diff
@@ -49,6 +49,14 @@ wf::input_method_relay::input_method_relay()
         auto evt_input_method = static_cast<wlr_input_method_v2*>(data);
         assert(evt_input_method == input_method);
 
+        // FIXME: workaround focus change while preediting
```

Comment:
This issue is if the user is preediting in one window, and switches to another, the preedit string will be "taken" over to the switched-over window.

> it seems we're likely to just send done() without anything else, is that a problem?

I don't see any problem yet. The application just discards the preedit string.

Reply:
> we will ignore the first commit after focus change

If the commit should belong to the new text input, it shouldn't have any meaningful content (e.g. empty).

> so that the next commit is basically empty

The problematic state is at the input method side, and we have no API to clear it. We can't notify the IM that the focus has changed or otherwise reset its state (input method v1 can, but wlroots doesn't support it).

Reply:
Hmm, so you mean that we may have some stale state in the IM client that it hasn't sent to wlroots yet? That sounds like a badly designed protocol but obviously we can't do much about it, I agree.

In this case, I'd at least like to have a much longer comment describing all of our discussion here, for future reference. Thanks for the explanations :)

Reply:
The state is maintained at the IM, not the compositor (like what will be committed if the user types what). We don't have the ability to tell the IM, "hey, the user has switched windows, please act accordingly."